okhttp/src/main/java/okhttp3/CertificatePinner.kt

```diff
@@ -240,21 +240,28 @@ class CertificatePinner internal constructor(
 
   internal data class Pin(
     /** A hostname like `example.com` or a pattern like `*.example.com`. */
-    val pattern: String,
+val pattern: String,
     /** The canonical hostname, i.e. `EXAMPLE.com` becomes `example.com`. */
-    private val canonicalHostname: String,
+private val canonicalPattern: String,
     /** Either `sha1/` or `sha256/`. */
-    val hashAlgorithm: String,
+val hashAlgorithm: String,
     /** The hash of the pinned certificate using [hashAlgorithm]. */
-    val hash: ByteString
+val hash: ByteString
   ) {
     fun matches(hostname: String): Boolean {
-      if (pattern.startsWith(WILDCARD)) {
-        val firstDot = hostname.indexOf('.')
-        return hostname.length - firstDot - 1 == canonicalHostname.length &&
-            hostname.startsWith(canonicalHostname, startIndex = firstDot + 1)
+      val patternSegments = canonicalPattern.split('.').reversed()
```

Comment:
I think avoiding this work when a simple hostname is used is still desirable.  Optimise for the == case.

Also Pin gives you a natural place to cache some of this work rather than computing each time.  Ultimately if a wildcard is in place, it may we worth checking endsWith before splitting anything.

Reply:
I have added a shortcut path for exact matches.